Add rounding support for `Periods`

spurll opened the pull request:
Allows (most) subtypes of `Period` to be rounded with `floor`, `ceil`, and `round`, also allowing for easy conversion between `Period` subtypes where we would otherwise encounter an `InexactError`.

```julia
julia> x = Dates.Millisecond(29394943)
29394943 milliseconds

julia> convert(Dates.Hour, x)
ERROR: InexactError: divexact(Int64, 8.165261944444444)
Stacktrace:
 [1] divexact(::Int64, ::Int64) at ./dates/periods.jl:400
 [2] convert(::Type{Base.Dates.Hour}, ::Base.Dates.Millisecond) at ./dates/periods.jl:422

julia> round(x, Dates.Hour)
8 hours
```

As with the date rounding functions, this approach also supports rounding to values (e.g., the nearest 15 minutes) as well as types:

```julia
julia> x = Dates.Hour(910)
910 hours

julia> ceil(x, Dates.Hour(12))
912 hours
```

I can see at least two potential concerns being raised about this approach:
1. As discussed in https://github.com/JuliaLang/julia/issues/18574, there may be some confusion/contention surrounding the typical rounding signature (`round(value, precision)`) and the convert signature (`convert(type, value)`).
2. ~For certain very large values, the rounding may still result in an `InexactError`. This is because all values are converted to `Nanosecond` for rounding, so we can't easily accommodate values that exceed `typemax(Int64)` nanoseconds (which is a little over 292 years). This could be mitigated by using a lower-precision `Period` type when nanosecond precision is not required.~

**Edit:** Point 2 above has been mitigated via `promote`.

spurll committed:
Add rounding for Periods

iamed2 commented:
Could/should these be `round(Dates.Hour, x)` and `round(x, Dates.Hour(12))` to match the `round(type, value)` and `round(value, precision)` methods? You could also conceivably have `round(Dates.Minute, x, Dates.Hour(12))` though I'm not sure how useful that would be.

spurll commented:
Unless something has changed, these signature actually do match the signatures of the date rounding functions, but the confusion is completely understandable (this is actually addressed in the first point of contention above).

spurll commented:
To clarify, both the "round a date(time) to a period value" and "round a date(time) to a period type" have the same signature, which is `round(val::TimeType, precision::Period)`, so this uses the same interface.

**But** I could definitely see an argument for "round to a period value" and "round to a period type" having different interfaces (with the latter using the same order as `convert`, as you suggest). I'm not sure whether that would lead to more or less confusion.

spurll commented:
The initial decision to use the rounding signature we currently have for date(time)s was informed by `trunc`, which already existed at the time. But to add to the confusion, `trunc` has a kind of inconsistent signature.

Generally, we have `trunc(T::Type, x)`. But for `TimeType`s we have   trunc(dt::TimeType, ::Type{Period})`.

iamed2 commented:
> But I could definitely see an argument for "round to a period value" and "round to a period type" having different interfaces (with the latter using the same order as convert, as you suggest). I'm not sure whether that would lead to more or less confusion.

This is the argument I was trying to make 😄

spurll commented:
I could definitely go either way on that one. But if we move to:
* `round(precision::Type{Period}, x)`
* `round(x, precision::Period`
then the interfaces for both `trunc` and the existing date rounding functions should be updated accordingly.

iamed2 commented:
We should see what @omus thinks

omus commented:
I think `round` for `Period` reminds me more of the `mod` behaviour rather than just indicating the return type:

```julia
julia> mod(260,256)
4

julia> mod(260,UInt8)
0x04
```

spurll commented:
That's interesting. So you're in favour of keeping the existing signatures, @omus?

omus commented:
Yes, I think the existing signatures are reasonable

omus commented on base/dates/rounding.jl:
```diff
@@ -160,21 +231,54 @@ function Base.round(dt::TimeType, p::Period, r::RoundingMode{:NearestTiesUp})
 # Make rounding functions callable using Period types in addition to values.
-Base.floor(dt::TimeType, p::Type{<:Period}) = Base.floor(dt, p(1))
-Base.ceil(dt::TimeType, p::Type{<:Period}) = Base.ceil(dt, p(1))
+Base.floor(x::TimeTypePeriod, p::Type{<:Period}) = Base.floor(x, p(1))
```
Should probably be changed to:
```julia
Base.floor(x::TimeTypePeriod, ::Type{P}) where P <: Period = Base.floor(x, oneunit(P))
```

spurll replied:
Interesting. I was completely unaware of `oneunit`.

omus replied:
It's the unit aware sibling to `one`

omus commented on test/dates/rounding.jl:
```diff
@@ -144,3 +144,69 @@ end
+end
+@testset "Rounding for periods that should not need rounding" begin
+    for x in [Dates.Week(3), Dates.Day(14), Dates.Microsecond(604800000000)]
+        local dt
```
Unused

omus commented on test/dates/rounding.jl:
```diff
@@ -144,3 +144,69 @@ end
+    for x in [Dates.Week(3), Dates.Day(14), Dates.Microsecond(604800000000)]
+        local dt
+        for p in [Dates.Week, Dates.Day, Dates.Hour, Dates.Second, Dates.Millisecond, Dates.Microsecond, Dates.Nanosecond]
+            local p
```
Unneeded?

omus replied:
@kshyatt can you elaborate on why this is needed?

spurll replied:
This is a case where I noticed that `local dt` had been added to the tests in several places, so I replicated that syntax for these tests (in this case `local dt` should actually be `local x` above). But yeah, I don't actually know what those lines are meant to accomplish.

omus commented on base/dates/rounding.jl:
```diff
@@ -7,6 +7,9 @@ const DATETIMEEPOCH = value(DateTime(0))
 const WEEKEPOCH = value(Date(0, 1, 3))
 
+const ConvertiblePeriod = Union{TimePeriod, Week, Day}
+const TimeTypePeriod = Union{TimeType, ConvertiblePeriod}
```
Maybe rename to `TimeTypeOrPeriod`

spurll committed:
Improve period rounding function type signatures

omus reviewed (approved):
Looks good. I'd like to know what the `local` loop variables are for yet.

omus commented:
Maybe spoke too soon @spurll there looks like there is a 32-bit issue.

spurll commented:
Ah yes. It looks like `Nanosecond` is using `Int` under the hood rather than `Int64`, so I'll have to address that.

spurll committed:
Use promote instead of Nanosecond to round periods

Period rounding functions no longer unnecessarily convert all values to
nanoseconds prior to rounding.

spurll commented:
AppVeyor failure seems unrelated. I believe everything else is fixed.

omus commented on base/dates/rounding.jl:
```diff
@@ -77,6 +80,35 @@ function Base.floor(dt::DateTime, p::TimePeriod)
+"""
+function Base.floor(x::ConvertiblePeriod, precision::T) where T <: ConvertiblePeriod
+    value(precision) < 1 && throw(DomainError(precision))
+    x, precision = promote(x, precision)
```
Should use a different variable name for the assignment to ensure type stability

omus commented on base/dates/rounding.jl:
```diff
@@ -160,21 +231,55 @@ function Base.round(dt::TimeType, p::Period, r::RoundingMode{:NearestTiesUp})
+function Base.round(x::ConvertiblePeriod, precision::ConvertiblePeriod, r::RoundingMode{:NearestTiesUp})
+    f, c = floorceil(x, precision)
+    common_x, common_f, common_c = promote(x, f, c)
+    return (common_x - common_f) < (common_c - common_x) ? f : c
```
I think `_x, _f, _c = ...` is fine if you want to use a more concise name

omus commented on base/dates/rounding.jl:
```diff
@@ -160,21 +231,55 @@ function Base.round(dt::TimeType, p::Period, r::RoundingMode{:NearestTiesUp})
-function Base.round(dt::TimeType, p::Type{<:Period}, r::RoundingMode=RoundNearestTiesUp)
-    return Base.round(dt, p(1), r)
+function Base.round(x::TimeTypeOrPeriod, ::Type{P}, r::RoundingMode=RoundNearestTiesUp) where P <: Period
+     return Base.round(x, oneunit(P), r)
```
Extra space

spurll committed:
Minor refactoring for period rounding

spurll commented:
I'm not sure what's going on with CI here.

omus commented:
Looks good. I'm going to retry the CI.
